### PR TITLE
TBPROTO: HTTPSTB: renego polishing

### PR DESCRIPTION
**Pull request by equalsJeffH**
this PR encompases some modest polishing of the recently-added TLS renegotiation language in TBPROTO & HTTPSTB, and clarification of the splitting of TBM across renego boundaries, plus some other minor editorial cleanups.

### CODE REVIEW

File: draft-ietf-tokbind-protocol-12.xml [hunk 1]
```diff
@@ -437,20 +437,23 @@
...
+        TLS master secret for the same connection, with associated session keys and EKM value.
+        TokenBinding.signature MUST be a signature of the EKM value derived from the TLS master
+        secret that produced the session keys encrypting the TLS application_data record containing
+        this TokenBinding. Such use of the current EKM for the TLS connection makes replay of bound
+        tokens within renegotiated TLS sessions detectable, but requires the application to
+        synchronize Token Binding message generation and verification with the TLS handshake
+        state.</t>
+
+        <t>Specifications defining the use of Token Binding with application protocols, such as
+        Token Binding over HTTP <xref target="I-D.ietf-tokbind-https"/>, may prohibit the use of TLS
```
Comment (martinthomson) [inline draft-ietf-tokbind-protocol-12.xml]: MAY

File: draft-ietf-tokbind-protocol-12.xml [hunk 2]
```diff
@@ -437,20 +437,23 @@
...
+        renegotiation in combination with Token Binding, obviating the need for such
+        synchronization. Alternatively, such specifications need to define a way to determine which
+        EKM value corresponds to a given TokenBindingMessage, and a mechanism preventing a
+        TokenBindingMessage from being split across TLS renegotiation boundaries (i.e., due to TLS
+        record layer fragmentation of "client messages" -- see Section 6.2.1 of <xref
+        target="RFC5246"/>).</t>
```
Comment (martinthomson) [inline draft-ietf-tokbind-protocol-12.xml]: I would also note that application-layer messages that carry a token binding might cross the boundary in ways that make processing difficult.  I don't think that we have generic guidance, but your HTTP proposal addresses that concern neatly.

File: draft-ietf-tokbind-protocol-12.xml [hunk 3]
```diff
@@ -353,9 +353,9 @@
   struct {
       TokenBindingType tokenbinding_type;
       TokenBindingID tokenbindingid;
-      opaque signature<0..2^16-1>;  /* Signature over the concatenation of 
-                                       tokenbinding_type, key_parameters and 
-                                       exported keying material (EKM) */
+      opaque signature<0..2^16-1>; /* Signature over the concatenation
+                                    of tokenbinding_type, key_parameters
+                                    and exported keying material (EKM)*/
```
Comment (martinthomson) [inline draft-ietf-tokbind-protocol-12.xml:358]: You didn't need to reflow this in this PR.

**Review by martinthomson — APPROVED** [review]

File: draft-ietf-tokbind-protocol-12.xml [hunk 4]
```diff
@@ -437,20 +437,23 @@
...
-        such as Token Binding over HTTP <xref target="I-D.ietf-tokbind-https"/>, may prohibit the 
-        use of TLS renegotiation in combination with Token Binding. Alternatively, such 
-        specifications need to define a way to determine which EKM value corresponds to a given 
-        TokenBindingMessage, and a mechanism preventing a TokenBindingMessage from being split 
-        across the renegotiation boundary.</t>
+        <t>Please note that TLS 1.2 and earlier versions support renegotiation, which produces a new
+        TLS master secret for the same connection, with associated session keys and EKM value.
+        TokenBinding.signature MUST be a signature of the EKM value derived from the TLS master
+        secret that produced the session keys encrypting the TLS application_data record containing
+        this TokenBinding. Such use of the current EKM for the TLS connection makes replay of bound
```
Comment (nharper) [inline draft-ietf-tokbind-protocol-12.xml]: (I should have noticed this in the last revision of this language.) The language "the TLS application_data record containing this TokenBinding" is problematic. The first problem is that a TokenBinding could be in multiple application_data records. This is fairly easily fixed by changing it to "application_data records", but could also need some language to say that all of those application_data records are encrypted with the same session keys.

The particularly problematic part is that there's no definition of what it means for an application_data record to contain a TokenBinding. An application protocol could be compressing its application data (including any TokenBindings), and the bits needed to reassemble a TokenBinding could be scattered across many non-contiguous application_data records. Do all of those application_data records count for this definition, or just the application_data record(s) that have the compressed form (which is some number of bits much less than the structure defined above)? If the latter, how does this spec express that idea, given that the application protocol could be encoding a TokenBinding however it wishes?

How does this language sound (in place of the sentence starting on line 442)?
"Any TokenBindingMessage sent after a TLS renegotiation MUST use for the TokenBinding.signature the EKM value derived from the TLS master secret that resulted from that renegotiation."

Comment (Andrei-Popov) [reply draft-ietf-tokbind-protocol-12.xml]: It seems sufficient to say "that produced the session keys encrypting **all** TLS application_data record**s** containing  this TokenBinding".

The problems of compression etc. are application-protocol specific, and we're deferring to the application protocol spec for these.

> How does this language sound (in place of the sentence starting on line 442)?
It seems to say the same thing in a slightly less concrete way.

Comment (Andrei-Popov) [inline draft-ietf-tokbind-protocol-12.xml]: It seems sufficient to say "that produced the session keys encrypting **all** TLS application_data record**s** containing  this TokenBinding".

The problems of compression etc. are application-protocol specific, and we're deferring to the application protocol spec for these.

> How does this language sound (in place of the sentence starting on line 442)?
It seems to say the same thing in a slightly less concrete way.

Comment (nharper) [reply draft-ietf-tokbind-protocol-12.xml]: What does it mean for an application_data record to contain a TokenBinding?

File: draft-ietf-tokbind-protocol-12.xml (hunk 3 above)
Comment (equalsJeffH) [reply draft-ietf-tokbind-protocol-12.xml:358]: those lines were too long.

File: draft-ietf-tokbind-protocol-12.xml (hunk 4 above)
Comment (Andrei-Popov) [reply draft-ietf-tokbind-protocol-12.xml]: We've agreed to correct this and say "records". A plurality of application_data records contain a Token Binding message when they contain bytes pertaining to the encoding of a Token Binding message? This seems pretty clear.

Comment (equalsJeffH) [reply draft-ietf-tokbind-protocol-12.xml]: Perhaps @martinthomson 's suggestion  https://github.com/TokenBinding/Internet-Drafts/pull/83#discussion_r91186993 overall addresses both @Andrei-Popov 's & @nharper 's concerns?

Comment (nharper) [reply draft-ietf-tokbind-protocol-12.xml]: Ok, that's fine. That definition of containing a TokenBinding means that if the application protocol does any sort of compression and doesn't reset the dictionary on TLS renegotiation, then whether or not a TokenBinding can be sent depends on whether the compression refers to a part of the dictionary before renegotiation. I don't know if this subtlety is intended.

Comment (Andrei-Popov) [reply draft-ietf-tokbind-protocol-12.xml]: I think the subtleties that pertain to HTTP are adequately addressed in HTTPSTB, and I agree with Martin's #83.

File: draft-ietf-tokbind-protocol-12.xml (hunk 1 above)
Comment (equalsJeffH) [reply draft-ietf-tokbind-protocol-12.xml]: thx.

File: draft-ietf-tokbind-protocol-12.xml (hunk 4 above)
Comment (equalsJeffH) [reply draft-ietf-tokbind-protocol-12.xml]: please check the new commits

File: draft-ietf-tokbind-protocol-12.xml (hunk 2 above)
Comment (equalsJeffH) [reply draft-ietf-tokbind-protocol-12.xml]: done.

**Pull request merged**